Update new about compression (#3012)
timescale/docs: use-timescale/compression/about-compression.md

<CompressionIntro />

This section explains how to enable native compression, and then goes into
detail on the most important settings for compression, to help you get the
best possible compression ratio.

## Key aspects of compression

> Compression always starts with the hypertable. Every table has a different schema but they do share some commonalities that we need to think about.

Reviewer: ```suggestion
Every table has a different schema but they do share some commonalities that you need to think about.
```

> Let's take the following schema as an example table named `metrics`:

Reviewer: ```suggestion
Consider the table `metrics` with the following attributes:
```

|Column|Type|Collation|Nullable|Default|
|-|-|-|-|-|
 time|timestamp with time zone|| not null|
 device_id| integer|| not null|
 device_type| integer|| not null|
 cpu| double precision|||
 disk_io| double precision|||

> Our hypertable needs to have a primary dimension, which you create when [creating a hypertable][create-hypertable]. In the example, we are showing the classic time-series use case with the `time` column as the primary dimension that is used for partitioning. Besides that we have two columns `cpu` and `disk_io` which are essentially value columns that we are capturing over time. There is also another column `device_id` which is used as a designator or a lookup key which designates which device these captured values belong to at a certain point in time.

Reviewer: ```suggestion
All hypertables have a primary dimension which is used to partition the table into chunks. The primary dimension is given when [the hypertable is created][create-hypertable]. In the example below, you can see a classic time-series use case with a `time` column as the primary dimension. In addition, there are two columns `cpu` and `disk_io` containing the values  that are captured over time and a column `device_id` for the device that captured the values.
```

Columns can be used in a few different ways:
You can use values in a column as a lookup key, in the example above the device_id is a typical example of such a column.
You can use a column for partitioning a table. This is typically a time column, but it is possible to partition the table using other columns as well.
You can use a column as a filter to narrow down on what data you select. The column device_type is an example of such a column where you can decide to only look at, for example, solid state drives (SSDs)

> The remaining columns typically are the values or metrics you are querying for. They are typically aggregated or presented in other ways. The columns `cpu` and `disk_io` are typical examples of such columns.

Reviewer: Very good way to structure the explanation. Here are a few minor suggestions to make it clearer, basically turning it into a bulleted list:
```suggestion
Columns can be used in a few different ways:

- You can use values in a column as a lookup key, in the example above `device_id` is a typical example of such a column.

- You can use a column for partitioning a table. This is typically a time column like `time` in the example above, but it is possible to partition the table using other types as well.

- You can use a column as a filter to narrow down on what data you select. The column `device_type` is an example of where you can decide to look at, for example, only solid state drives (SSDs).

The remaining columns are typically the values or metrics you are collecting. These are typically aggregated or presented in other ways. The columns `cpu` and `disk_io` are typical examples of such columns.
```

`} />

When chunks are compressed in a hypertable, data stored in them is reorganized and stored in column-order rather than row-order. As a result, it is not possible to use the same uncompressed schema version of the chunk and a different schema must be created. This is automatically handled by TimescaleDB, but it has a few implications:
The compression ratio and query performance is very dependent on the order and structure of the compressed data, so some considerations are needed when setting up compression.
Indexes on the hypertable cannot always be used in the same manner for the compressed data.

<Highlight type="note">
Indexes set on the hypertable are used only on chunks containing uncompressed
data. Timescale creates and uses custom indexes to incorporate the `segmentby` 
and `orderby` parameters during compression which are used when reading compressed data.
More on this in the next section.
</Highlight>

Based on the previous schema, filtering of data should happen over a certain time period and analytics are done on device granularity. This pattern of data access lends itself to organizing the data layout suitable for compression.

### Segmenting and ordering.

> Segmenting the compressed data should be based on the way you access the data. Basically, you want to segment your data in such a way that you can make it easier for your queries to fetch the right data at the right time. That is to say, your queries should dictate how you segment the data so they can be optimized and yield even better query performance.

Reviewer 2: This is a little confusing because it starts using the term "segmenting" before explaining what this is. Probably it should go like this:
1) why you have to order data for compression
2) how you do this in Timescale (orderby column)
3) but we also have a segmentby and it's useful in the following ways

By the way, why do we use segmentby at all? The only thing that comes to mind is batch sorted merge, i.e. you can have efficient point queries on segmentby, and at the same time efficient ordering by orderby. Does it have other uses? For point queries on segmentby, I think the sparse index would be good enough.

Author: I completely agree with this ordering, in fact, I did it that way before but got feedback to start with segmenting first.

Now I'm not completely sure what to do.

> By the way, why do we use segmentby at all? The only thing that comes to mind is batch sorted merge, i.e. you can have efficient point queries on segmentby, and at the same time efficient ordering by orderby. Does it have other uses? For point queries on segmentby, I think the sparse index would be good enough.

If you were to get rid of segmentby, would you move those columns to orderby or no? Grouping data by those columns feels like a natural ordering but if you move them to orderby I could see issues with adding new data to already compressed chunks. It would necessitate a lot more decompression to target specific segment groups. Thats the first thought that comes to mind.

Reviewer 2: Yes, if we're not using segmentby, that column should go to orderby. Good point about decompression granularity, but minmax indexes could help there, not sure if we're using them now.

Author: Changed the order of ordering and segmenting sections.

`} />

When chunks are compressed in a hypertable, data stored in them is reorganized and stored in column-order rather than row-order. As a result, it is not possible to use the same uncompressed schema version of the chunk and a different schema must be created. This is automatically handled by TimescaleDB, but it has a few implications:
The compression ratio and query performance is very dependent on the order and structure of the compressed data, so some considerations are needed when setting up compression.
Indexes on the hypertable cannot always be used in the same manner for the compressed data.

<Highlight type="note">
Indexes set on the hypertable are used only on chunks containing uncompressed
data. Timescale creates and uses custom indexes to incorporate the `segmentby` 
and `orderby` parameters during compression which are used when reading compressed data.
More on this in the next section.
</Highlight>

Based on the previous schema, filtering of data should happen over a certain time period and analytics are done on device granularity. This pattern of data access lends itself to organizing the data layout suitable for compression.

### Ordering and segmenting.

> Ordering the data will have a great impact on the compression ratio since we want to have rows that change over a dimension (most likely time) close to each other. Most of the time data changes in a predictable fashion, following a certain trend. We can exploit this fact to encode the data so it takes less space to store. For example, if you order the records over time, they will get compressed in that order and subsequently also accessed in the same order.

Reviewer: It affects both compression ratio and performance.

> This makes the time column a perfect candidate for ordering your data since the measurements evolve as time goes on. If you were to use that as your only compression setting, you would most likely get a good enough compression ratio to save a lot of storage. However, accessing the data effectively depends on your use case and your queries. With this setup, you would always have to access the data by using the time dimension and subsequently filter all the rows based on any other criteria.

Reviewer: The interleaved caveat makes the text a little hard to read. Would suggest to just start with the "ordering has a great impact", "time is usually a good column for this", "this is the command to get that", "this is the result", discussion about "it is not always this simple".

I usually do it in the order: setting the stage by giving context, follow with concrete things and examples, end with elaborations and caveats.

`time` column is used for ordering data, which makes filtering it using `time` column much more efficient.

<CodeBlock canCopy={false} showLineNumbers={false} children={`
postgres=# select avg(cpu) from metrics where time >= '2024-03-01 00:00:00+01' and time < '2024-03-02 00:00:00+01';
        avg         
--------------------
 0.4996848437842719
(1 row)
Time: 87,218 ms
postgres=# ALTER TABLE metrics 
SET (
	timescaledb.compress, 
	timescaledb.compress_segmentby = 'device_id',
	timescaledb.compress_orderby='time'
);
ALTER TABLE
Time: 6,607 ms
postgres=# SELECT compress_chunk(c) FROM show_chunks('metrics') c;
             compress_chunk             
----------------------------------------
 _timescaledb_internal._hyper_2_4_chunk
 _timescaledb_internal._hyper_2_5_chunk
 _timescaledb_internal._hyper_2_6_chunk
(3 rows)
Time: 3070,626 ms (00:03,071)
postgres=# select avg(cpu) from metrics where time >= '2024-03-01 00:00:00+01' and time < '2024-03-02 00:00:00+01';
       avg        
------------------
 0.49968484378427
(1 row)
Time: 45,384 ms
`} />

> Segmenting the compressed data should be based on the way you access the data. Basically, you want to segment your data in such a way that you can make it easier for your queries to fetch the right data at the right time. That is to say, your queries should dictate how you segment the data so they can be optimized and yield even better query performance.

Reviewer 2: This sounds very abstract, can we say something more specific here? E.g. the columns on which you have equality filters should go to segmentby. But only if you have > 100 values with same segmentby, otherwise the compression ratio will be bad, and the column should better go to orderby.

Reviewer 2: I think that when we start describing the segmentby columns, we can do it from the top down, e.g.:
- for columns that often have an equality filters, it is possible to keep them uncompressed
- this allows you to use the normal PG index on the compressed table to satisfy these filters
- this also allows you to granularly update the compressed data, not decompressing the other segments
- and this allows to order on the "orderby" fast, i.e. batch sorted merge.
- but the condition is that you should have at least 100 values per segmentby, otherwise there will be data bloat.
So normally for the time series we use sensor id as a segmentby, and time column as the first orderby column. This way getting the last entries is fast, and getting the particular sensor is fast, and the data compresses reasonably well.

I think if we add some text along these lines, it will be more clear.

Author: Only problem with this might be that it goes a lot in depth for being an "about compression" page. 

It would make sense to have it in a follow up page like "compression design".

WDYT?

Reviewer 2: Yeah, maybe we can split it. But I think it's important to give a good explanation of how segmentby is different from orderby, so that people can decide what they need to use. Especially given that segmentby is somewhat unique to Timescale.

Reviewer: It might make sense to focus on a few queries where segmentby is good rather than trying to create an all-encompassing description... something like this:

1. Consider this query
2. Here is the plan
3. Here is why it would be better to use a segment by for this
4. Rinse and repeat a few times.

Using the following configuration setup on our example table:
<CodeBlock canCopy={false} showLineNumbers={false} children={`
ALTER TABLE metrics 
SET (
	timescaledb.compress, 
	timescaledb.compress_segmentby='device_id', 
	timescaledb.compress_orderby='time'
);
`} />

would produce the following data layout.

|time|device_id|device_type|cpu|disk_io|energy_consumption|
|---|---|---|---|---|---|
|[12:00:02, 12:00:01]|1|[SSD,SSD]|[88.2, 88.6]|[20, 25]|[0.8, 0.85]|
|[12:00:02, 12:00:01]|2|[HDD,HDD]|[300.5, 299.1]|[30, 40]|[0.9, 0.95]|
|...|...|...|...|...|...|

<Highlight type="note">
Number of rows that are compressed together in a single batch (like the ones we see above) is 1000.
If your chunk does not contain enough data to create big enough batches, your compression ratio will be reduced.
This needs to be taken into account when defining your compression settings.
</Highlight> 

> 

Reviewer: This probably makes more sense in a dedicated "caveat" section.

Author: Agree... makes more sense in compression design but I'll leave it here for the time being, until we get around to revamping that page.
